src/SparkFun_AS7331.h

```diff
@@ -14,1312 +14,716 @@
+    bool _standbyState;         // Local state value. False means the device is not in standby mode.
+    bool _startState;           // Local state value. False means the device is not allowed to measure.
+    bool _powerDownEnableState; // Local state value. False means the device is NOT POWERED DOWN. This is inverse.
+
+    as7331_dev_op_state_t _opMode; // Local state value. Configuration or Measure operating mode.
+    as7331_gain_t _sensorGain;     // Local config value. Sensor gain stored as (1 << (11 - gain)).
+    as7331_conv_clk_freq_t _cclk;  // Local config value. Sensor's conversion clock stored as 1024*(1 << cclk) kHz.
+    as7331_meas_mode_t _mmode; // Local state value. Details the device's measurement mode as CONT, CMD, SYNS, or SYND.
+    as7331_conv_time_t _conversionTime; // Local config value. Contains the conversion time stored as (1 << time) ms.
+    as7331_divider_val_t _dividerRange; // Local config value. Contains the predivider value stored as (1 << 1 + range).
+
+    float _uva;         // Last valid UVA result in uW/cm2.
+    float _uvb;         // Last valid UVB result in uW/cm2.
+    float _uvc;         // Last valid UVC result in uW/cm2.
+    float _temperature; // Last valid temperature result in degC.
+
+    uint32_t _outputConversionTime; // Last valid output conversion time result in of clock cycle count.
+
+    float _conversionA; // Internal conversion factor for the UVA sensor.
+    float _conversionB; // Internal conversion factor for the UVB sensor.
+    float _conversionC; // Internal conversion factor for the UVC sensor.
+
+    const float _fsrA = 348160.0; // Full Scale Resolution for the UVA sensor.
+    const float _fsrB = 387072.0; // Full Scale Resolution for the UVB sensor.
+    const float _fsrC = 169984.0; // Full Scale Resolution for the UVC sensor.
+};
 
-            sfe_as7331_reg_cfg_optreg_t optreg;
+class SfeAS7331ArdI2C : public SfeAS7331Driver
+{
+  public:
+    SfeAS7331ArdI2C()
+    {
+    }
 
-            result = getOptIndex(&optreg);
-            if(SFE_BUS_OK != result)
-                return result;
+    /// @brief  Sets up Arduino I2C driver using the specified I2C address then calls the super class begin.
+    /// @param address Address of the I2C device.
+    /// @return True if successful, false otherwise.
+    bool begin(const uint8_t &address = kDefaultAS7331Addr)
+    {
+        if (_theI2CBus.init(address) != kSTkErrOk)
+            return false;
 
-            optreg.init_idx = indexMode;
+        // Device supports repeat starts, enable it.
+        _theI2CBus.setStop(false);
 
-            result = setOptIndex(&optreg);
-            if(SFE_BUS_OK != result)
-                return result;
+        setCommunicationBus(&_theI2CBus);
 
-            _config.indexMode = indexMode;
+        if (!reset())
+            return false;
 
-            return SFE_BUS_OK;
-        }
+        isConnected();
```

Comment:
Going back to comments from the previous implementation, you can't make assumptions about the operating mode of the device on reset, so you need to make sure it's in the right mode. The easiest way to do this is to reset the device to make sure it's in a known configuration. This is standard practice.

Reply:
I understand that the operating mode needs to be correct to do stuff, but that's not my point. The point is that resetting the device and checking if it's connected (whatever order that happens in) is independent of what platform you're on, so they should be done within the base class to reduce duplicate code and keep the scope of things where they should be.

To your other point about `isConnected()` being implementation-specific - just the act of calling `getDeviceID()` effectively also pings the device, because if it's not on the bus, then that creates an error that gets handled differently. But if you want to be "proper" about it, then you could add a virtual `isConnected()` in the base class, and implement that in the derived classes. Then you can still move all the code for resetting and checking if it's connected into the base class, which will reduce duplicate code and keep the scope of things where they should be.

Reply:
I removed the reset, so this isn't really valid anymore. Now that getDeviceID will move it to cfg mode and back it wasn't necessary. 

However, getting the deviceID is different from ping and is interface specific.

I don't agree that by adding a second isConnected and doing half in there and half in the implementation specific would in any way reduce duplicate code. If anything it would increase it.